readme.md
<p align="center">
	<a href="#">
		<img width="300" src="https://github.com/dannysmc95/rewyre/blob/master/rewyre-logo.svg">
	</a>
</p>

<p align="center">
	Rewyre is a fast and flexible wiring tool for developing RESTful and WebSocket APIs using TypeScript decorators.
</p>

<p align="center">
	<img src="https://img.shields.io/npm/v/rewyre?color=blue">
	<img src="https://img.shields.io/npm/l/rewyre">
	<img src="https://img.shields.io/npm/dw/rewyre">
</p>

The rewyre framework is built on top of express, and express-ws and utilises TypeScript and decorators to give you a powerful wiring tool to create API applications in a very short time with built-in support for MySQL/MongoDB for models.


<br />


<a target="_blank" href="https://discord.gg/3YuNTEMJey">
	<p>
		<img style="display: inline-block; position: relative; top: 5px;" width="25" src="https://i.redd.it/s9biyhs4lix61.jpg" />
		<span style="padding-left: 10px; font-size: 20px;">Have a question or suggestion? Join the discord!</span>
	</p>
</a>


<br />


## Important Changes

> See the docs/migrating/v1-v2.md for more information regarding the move to V2.

> The framework's version 2 release is now licensed under MIT.


<br />


## Installation

You can install from NPM with:

```plaintext
npm install --save rewyre
```


<br />


## Bootstrap Application

As part of some new changes, I have added a simple bootstrap script, it doesn't do very much other than create a basic application with eslint, and all the required packages, you can use it like:

```
npx rewyre create-rewyre <foldername/appname>
```


<br />


## Getting Started

Below is a simple example showing a database connection and a simple find, for a more in-depth example, have a look at the `test` folder in the source, which has a simple to-do demo.

**Note**: _The expectation is that this framework is to be used with TypeScript specifically and that you would you use your own tooling for building and compiling it, of course, you can use JavaScript and babel instead, but it is suggested to use TypeScript, for help with setting up, look at the tsconfig.json file inside of the test folder._

```typescript
// Import the parts we need.
import { Framework, Drivers, Controller, Route, IReturn, Model, AbstractModel, AbstractController } from 'rewyre';


// Define our controller.
@Controller('/', 'home')
class HomeController extends AbstractController {

	@Route('GET', '/')
	public async index(): Promise<IReturn> {
		const users = await this.users.find({});
		return { status: 200, content: users };
	}
}


// Define our model.
@Model('users', 'general', {
	username: 'string',
	password: 'string',
	blocked: 'boolean',
	block_reason: '?string',
})
class UsersModel extends AbstractModel {}


// Create an isolated async function.
(async() => {

	// Create an instance of the framework.
	const application: Framework = new Framework({
		port: 3005,
		database: true,
		databases: [
			{
				unique: 'primary',
				host: 'localhost',
				port: '27017',
				name: 'example-app',
				driver: Drivers.MONGO,
				default: true,
			},
		],
	});

	// Register classes.
	application.register([
		HomeController,
		UsersModel,
	]);

	// Start the server.
	await application.start();
})();


```


<br />


## Available Features

The below lists the features and their stable state, this framework's API will not change for the forseable future, any changes will be fully implemented and any non-backwards compatible changes will be in the latest major version, following the semver versioning scheme.

| Feature | Description | Status |
| - | - | - |
| HTTP Server | The HTTP server is the base for the framework and is built on top of Express. | `Stable` |
| WebSocket Server | The WebSocket server uses `express-ws` package to apply WebSocket support. | `Stable` |
| Middleware Support | Standard Express middleware is supported using the `useMiddleware` method. | `Stable` |
| Static Assets | Standard Express static is supported as well using the `useStatic` method. | `Stable` |
| Controllers | Controller classes and the `@Controller` decorator are both implemented. | `Stable` |
| Controller Routes | Controller routes are and the `@Route` decorator are both implemented. | `Stable` |
| Models | Model classes and the `@Model` decorator are both implemented. | `Stable` |
| Drivers | Database drivers allow you to use one of the many implemented into the framework. | `Stable` |
| Custom Drivers | Define your own database drivers to implement other databases into your framework. | `Stable` |
| Framework Hooks | Allows you to hook functions into internal events. | `Beta` |
| Plugin Support | See documentation but plugins are now implemented to allow you to package code pieces into reusable and shareable components. | `Beta` |
| Multiple Databases | Your models can use any database, including multiple, have data in many databases? Write a model around a specific database instead or just fallback to the default. | `Beta` |
| Injections | Injections are done using a single `@Inject` decorator and you can inject one or many, you can inject models and providers to any service or controller as required. | `Stable` |
| Services | Service classes and the `@Service` decorator are both implemented and services can run on a loop based on seconds. | `Stable` |
| Providers | Providers and the `@Provide` decorator are both implemented, the provider allows you to create built in helper classes that can be injected to controllers, and services. | `Stable` |


<br />


## Upcoming Features

Upcoming features that are planned, some features may come out quicker due to technical requirements.

* ~~**Plugin Support** - There is a plan to add plugin support, but I am not sure what this will look like yet.~~
* ~~**Multiple Databases** - Support for using multiple databases using database drivers alongside your own.~~
* **Model Extend** - Allow developers to extend models to add more fields for validation.


<br />


## Future Plans

The future for this library is big, I have many plans to add lots of new features and more decorators to add additional features, and I plan to turn this (slowly) into a full out-of-the-box TypeScript based framework for building Node.JS server side applications. This framework will be able to handle any structure and can be included into other applications like Vue.JS Server-Side Rendering alongside using it render HTML based applications instead of only API based.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v2.5.0](https://github.com/dannysmc95/rewyre/compare/v2.4.4...v2.5.0) - 2021-09-20

### Commits

- Switched to using the Yarn package manager, bumped @types/node @types/mongodb and the mongodb packages to the latest version. [`a450daa`](https://github.com/dannysmc95/rewyre/commit/a450daa6267f74e7bb286c2fcd7ac6db0e36c90e)
- Supporting the latest version of MongoDB, and updated all packages to the latest version, manually re-exporting ObjectID as this has been deprecated in favour of ObjectId. [`4d984a9`](https://github.com/dannysmc95/rewyre/commit/4d984a93bc67f405bec88f615e5eb10a490c23d8)
- Finalised cron support, added a demo case to trial support. [`20630db`](https://github.com/dannysmc95/rewyre/commit/20630db7fd1c6e973fddae0c075369371ee23f87)
- Added notice about using the new CLI method. [`b824de0`](https://github.com/dannysmc95/rewyre/commit/b824de0497895a93f671869e90abd474ee9d1649)
- Cleaned up launch script. [`385ce26`](https://github.com/dannysmc95/rewyre/commit/385ce263d2dbb8243b7faa324c0cc78bb93c1de0)
- Added support for cron-based service scheduling. [`248d013`](https://github.com/dannysmc95/rewyre/commit/248d013eb7d27c5dbcac7a282b58780d6cd34a99)
- Update readme.md [`dc61eba`](https://github.com/dannysmc95/rewyre/commit/dc61eba56697184e568a8629ed9274659d3e9697)
- Update readme.md [`579a870`](https://github.com/dannysmc95/rewyre/commit/579a870923581d57dd50785e586e5ec5bb740b71)

## [v2.4.4](https://github.com/dannysmc95/rewyre/compare/v2.4.3...v2.4.4) - 2021-06-16

### Commits

- Added new bin command for bootstrapping a rewyre project. [`a328ccb`](https://github.com/dannysmc95/rewyre/commit/a328ccbea22294353217ef652335ed3c274f6794)
- Added new serverOptions that allows you to pass over options to the underlying websocket instance. [`e095ba8`](https://github.com/dannysmc95/rewyre/commit/e095ba80d405b9ee0cb818b0f17c00c0e0c0a596)
- Updated changelog. [`3ea735c`](https://github.com/dannysmc95/rewyre/commit/3ea735c7859052cb6672afb66cb940c932244da8)
- Bumped the package version. [`f35a7c1`](https://github.com/dannysmc95/rewyre/commit/f35a7c1a292bb24ca250c00b92b35dd14dca2076)
- Updated a file, as it was half finished. [`01e015e`](https://github.com/dannysmc95/rewyre/commit/01e015e74c6a6396f1bd073210dec20754826a8f)
- Update readme.md [`c26d887`](https://github.com/dannysmc95/rewyre/commit/c26d88788620c68100a8fe6c2e52b2141baee89f)

## [v2.4.3](https://github.com/dannysmc95/rewyre/compare/v2.4.2...v2.4.3) - 2021-06-11

### Commits

- Fixed lint errors with latest changes. [`7c28453`](https://github.com/dannysmc95/rewyre/commit/7c284539d521c6c131acdc53f6ebe508964b8993)
- Bumped version to 2.4.3 as I missed the lint errors. [`766a7ba`](https://github.com/dannysmc95/rewyre/commit/766a7ba70868a1795832011ea0b096b5a419a7eb)

## [v2.4.2](https://github.com/dannysmc95/rewyre/compare/v2.4.1...v2.4.2) - 2021-06-11

### Fixed

- Removed the database option from the state, as this needs to be re-looked at, instead changed the state type to 'file' or 'in-memory', fixes #8 [`#8`](https://github.com/dannysmc95/rewyre/issues/8)
- Added change that checks the build environment (whether using .ts or .js files) if .ts then process.env.BUILD_ENV is defined as dev or prod for .js, fixes #11 [`#11`](https://github.com/dannysmc95/rewyre/issues/11)

### Commits

- Added fix to the plugin options, to prevent them from being overwritten by the plugin's own defined settings so that users can override them. [`898cca2`](https://github.com/dannysmc95/rewyre/commit/898cca2a2c61d49fa6f3317f751081a91f21bc6d)
- Updated changelog. [`84f6f1d`](https://github.com/dannysmc95/rewyre/commit/84f6f1db0c66255a22e7d14485d914b531122bf1)
- Added discord link so people can join and discuss. [`e85a957`](https://github.com/dannysmc95/rewyre/commit/e85a9570f65d6fa69c9c3407521bc303366637d3)
- Update readme.md [`3b7c08c`](https://github.com/dannysmc95/rewyre/commit/3b7c08c5505c32309d8cd4c1f8cebe30c9ad37de)
- Updated readme [`4b55a85`](https://github.com/dannysmc95/rewyre/commit/4b55a85d7e0c595c327437518539b43133a4e5ca)
- Bumped patch version. [`2933f08`](https://github.com/dannysmc95/rewyre/commit/2933f084e76219688c941044ec50b6b00da8d5ae)

## [v2.4.1](https://github.com/dannysmc95/rewyre/compare/v2.4.0...v2.4.1) - 2021-06-07

### Commits

- Updated all dependencies and upgraded express-ws to the latest version to remove security issue. [`93c5675`](https://github.com/dannysmc95/rewyre/commit/93c5675ab331ef41c0b38a43c3d1aad600707693)
- Updated changelog. [`77db24b`](https://github.com/dannysmc95/rewyre/commit/77db24ba20c5d4b2f4a7bc34a76f1cb80aef57d5)
- Added new method to the framework called getLogger to allow hooks to access logging. [`9c48f18`](https://github.com/dannysmc95/rewyre/commit/9c48f18793d0d656e9f059ae3a22740e4d600b8c)
- Added small advisory relating to the database driver issue. [`ba86b3d`](https://github.com/dannysmc95/rewyre/commit/ba86b3dcc158475b6ccc145975543655cf2e1705)
- Bumped package version. [`15a0be6`](https://github.com/dannysmc95/rewyre/commit/15a0be6f7e01f7bec6aa3d8eb1858ed982a2ac82)

## [v2.4.0](https://github.com/dannysmc95/rewyre/compare/v2.3.0...v2.4.0) - 2021-06-07

### Commits

- Added new registry module for internal state management, alongside exporting it from the framework. [`ec461aa`](https://github.com/dannysmc95/rewyre/commit/ec461aa2bf4971b8ee2192c10c4557fdd21dca2c)
- Updated changelog. [`ca545d7`](https://github.com/dannysmc95/rewyre/commit/ca545d72ba3223d776200746e038b47f425c4782)
- Added new framework getRegistry method for accessing the registry, alongside setting it as a built-in injectable. [`a269ed7`](https://github.com/dannysmc95/rewyre/commit/a269ed72dd87ccfb0c4af4ae86365b78e986c34b)

script/cli.js
#!/usr/bin/env node

console.log('Rewyre CLI V0.1');

// Require libraries.
const originalPackage = require('../package.json');
const spawn = require('child_process').spawn;
const resolve = require('path').resolve;
const { mkdirSync, writeFileSync } = require('fs');

// Get the process args and package.
const rawname = process.argv.pop();
const name = rawname.indexOf('/usr/bin') === -1 ? rawname.replace(/[^\w\s-]/gi, '') : 'rewyre-project';

// Define the file structures.
const folderlist = ['src', 'src/controller'];
const filelist = {};

// Define our base package file.
filelist['package.json'] = `{
	"name": "${name}",
	"version": "0.1.0",
	"description": "A rewyre application.",
	"main": "dist/application.js",
	"scripts": {
		"build": "rm -rf lib/ && node_modules/.bin/tsc",
		"lint": "node_modules/.bin/eslint . --ext .ts",
		"dev": "ENV=dev node -r ts-node/register --max_old_space_size=2048 ./src/application.ts",
		"start": "ENV=prod node --max_old_space_size=2048 ./dist/application.js"
	},
	"author": "Someone <someone@somewhere.com> (https://somewhere.com/)",
	"license": "UNLICENSED",
	"dependencies": {
		"rewyre": "^${originalPackage.version}"
	},
	"devDependencies": {
		"@types/node": "^15.12.2",
		"@typescript-eslint/eslint-plugin": "^4.27.0",
		"@typescript-eslint/parser": "^4.27.0",
		"eslint": "^7.28.0",
		"ts-node": "^10.0.0",
		"typescript": "^4.3.2"
	}
}
`;

// Create our files.
filelist['eslintrc.js'] = `module.exports = {
	extends: [
		'eslint:recommended',
		'plugin:@typescript-eslint/recommended'
	],
	env: {
		node: true,
		es6: true,
		es2017: true,
		es2020: true,
		es2021: true
	},
	globals: {
		'__APP_VERSION__': true
	},
	parser: '@typescript-eslint/parser',
	parserOptions: {
		'ecmaVersion': 2020,
		'sourceType': 'module'
	},
	plugins: [
		'@typescript-eslint/eslint-plugin'
	],
	rules: {
		'@typescript-eslint/interface-name-prefix': 0,
		'@typescript-eslint/no-explicit-any': 0,
		'@typescript-eslint/no-this-alias': 0,
		'@typescript-eslint/no-var-requires': 0,
		'@typescript-eslint/explicit-module-boundary-types': 0,
		'@typescript-eslint/no-unused-vars': ['error', {
			'vars': 'all',
			'args': 'after-used',
			'ignoreRestSiblings': false
		}],
		'semi': ['error', 'always', { 'omitLastInOneLineBlock': true}],
		'no-useless-escape': 2,
		'handle-callback-err': 2,
		'no-fallthrough': 2,
		'no-new-require': 2,
		'max-len': [2, 240],
		'camelcase': 0,
		'require-jsdoc': 0,
		'valid-jsdoc': 0,
		'prefer-spread': 1,
		'prefer-rest-params': 1,
		'linebreak-style': 0,
		'quote-props':[0, 'as-needed'],
		'no-prototype-builtins': 0,
		'@typescript-eslint/no-useless-constructor': ['error'],
		'@typescript-eslint/explicit-member-accessibility': ['error'],
		'@typescript-eslint/camelcase': 0,
		'@typescript-eslint/ban-types': ['error', {
			'types': {
				'Function': false,
			},
			'extendDefaults' : true,
		}],
		'@typescript-eslint/naming-convention': ['error',
			{ selector: 'default', format: ['PascalCase', 'camelCase', 'snake_case', 'UPPER_CASE'] },
			{ selector: 'function', format: ['PascalCase', 'camelCase'], leadingUnderscore: 'forbid' },
			{ selector: 'variable', format: ['snake_case', 'camelCase', 'PascalCase'], leadingUnderscore: 'allow' },
			{ selector: 'memberLike', format: ['camelCase', 'PascalCase', 'UPPER_CASE', 'snake_case'], leadingUnderscore: 'allow' }
		],
		'no-trailing-spaces': ['error', {
			ignoreComments: true,
			skipBlankLines: false
		}],
		'space-before-function-paren': ['error', {
			anonymous: 'never',
			named: 'never',
			asyncArrow: 'always',
		}],
		'padded-blocks': 0,
		'indent': ['error', 'tab', {
			SwitchCase: 1,
		}],
		'no-tabs': 0,
		'no-useless-return': 0,
		'comma-dangle': ['error', {
			arrays: 'always-multiline',
			objects: 'always-multiline',
			imports: 'always-multiline',
			exports: 'always',
			functions: 'always-multiline',
		}],
		'quotes': ['error', 'single'],
		'keyword-spacing': ['error', {
			overrides: {
				catch: {
					before: true,
					after: false,
				},
			},
		}],
		'array-bracket-spacing': 0,
		'generator-star-spacing': 0,
		'one-var': 0,
		'no-mixed-operators': 0,
		'eol-last': ['error', 'always'],
	}
}
`;

filelist['.eslintignore'] = `node_modules
coverage/
lib/
.prettierrc
.vscode/
.circleci/
`;

filelist['tsconfig.json'] = `{
	"compilerOptions": {
		"target": "es6",
		"module": "commonjs",
		"outDir": "./dist",
		"rootDir": "./src",
		"declaration": true,
		"sourceMap": true,
		"noImplicitAny": false,
		"noImplicitThis": false,
		"moduleResolution": "node",
		"strict": true,
		"experimentalDecorators": true,
		"emitDecoratorMetadata": true,
		"removeComments": false,
		"resolveJsonModule": true,
		"types": ["node"],
		"typeRoots": [ "node_modules/@types" ]
	},
	"exclude": [ "test", "dist" ]
}
`;

filelist['readme.md'] = `# ${name}

A blank rewyre project.

To run your application in development mode:
\`\`\`
npm run dev
\`\`\`

To lint your application:
\`\`\`
npm run lint
\`\`\`

To build your application:
\`\`\`
npm run build
\`\`\`

To start your application in production mode:
\`\`\`
npm start
\`\`\`
`;

filelist['src/application.ts'] = `import { Framework, Drivers } from 'rewyre';
import { HomeController } from './controller/home';

(async () => {

	// Create an instance of the framework.
	const application: Framework = new Framework({
		port: 8080,
		database: true,
		log_levels: ['info', 'warn', 'error', 'verbose', 'debug'],
		databases: [
			{
				unique: 'main',
				host: 'localhost',
				port: 27017,
				name: '${name}-development',
				driver: Drivers.MONGO,
				default: true,
			},
		],
	});

	// Register classes.
	application.register([
		HomeController,
	]);

	// Start the server.
	await application.start();
})();
`;

filelist['src/controller/home.ts'] = `import { Controller, AbstractController, Route, IReturn } from 'rewyre';

@Controller('/', 'home')
export class HomeController extends AbstractController {

	@Route('GET', '/')
	public async index(): Promise<IReturn> {
		return { status: 200, content: 'Hello, World!' };
	}
}
`;

// Now create folders.
console.log('-> Creating directories...');
const baseFolder = process.cwd();
mkdirSync(resolve(baseFolder, `./${name}`));
folderlist.forEach(foldername => {
	mkdirSync(resolve(baseFolder, `./${name}`, `./${foldername}`));
});

// Now loop and create files.
console.log('-> Creating files...');
Object.keys(filelist).forEach(path => {
	writeFileSync(resolve(baseFolder, `./${name}`, `./${path}`), filelist[path], 'utf-8');
});

// Now run install.
console.log('-> Running NPM install...');
const npmInstall = spawn('npm', ['install'], { cwd: resolve(baseFolder, `./${name}`) });
npmInstall.stdout.on('data', data => { console.log(data.toString('utf-8')) });
npmInstall.stderr.on('data', data => { console.error(data.toString('utf-8')) });
npmInstall.on('exit', code => {

	// On exit.
	if (code.toString() !== '0') {
		console.error(`Process exited with code ${code.toString()}.`);
	}

	// Now notify it's done.
	if (code.toString() === '0') {
		console.log('Setup was completed.');
	}
});
